enterprise/server/remote_execution/workspace/workspace.go

```diff
@@ -270,8 +298,23 @@ func (ws *Workspace) UploadOutputs(ctx context.Context, cmd *repb.Command, execu
 		return nil
 	})
 	eg.Go(func() error {
+		if ws.overlay != nil {
+			// When overlayfs is enabled, apply the changes from upperdir to
+			// lowerdir, since dirHelper's root dir is configured as the
+			// lowerdir and it needs to see the output files.
+			//
+			// Optimization: if recycling is not enabled, then at this point the
+			// runner should be removed and cannot affect any files in the
+			// workspace anymore, so it is safe to rename the outputs files in
+			// upperdir here rather than copying.
+			recyclingEnabled := platform.IsTrue(platform.FindValue(ws.task.GetCommand().GetPlatform(), platform.RecycleRunnerPropertyName))
```

Comment:
Without this optimization, we have to do full copies of all output files in the action instead of being able to hardlink like we do today. I think the IO savings will be worth the small amount of extra complexity.

The comment here frames it as an optimization but when I was originally writing the code I thought of it more as handling an edge case - hardlinking is our preferred/simplest way of copying files, but when runner recycling is enabled we have to fall back to a clunky/inefficient full-copy for safety reasons.

Reply:
ok.